Sync HardwareIsolation 1030 commits to 1050

deepakala-k opened the pull request:
Sync HardwareIsolation commits from 1030 to 1050. 
Here is the upstream commits for the relevant ones. 
Implement LocationIndicatorActive for CPU resource | https://gerrit.openbmc.org/c/openbmc/bmcweb/+/37045
Implement SubProcessors for processor core | https://gerrit.openbmc.org/c/openbmc/bmcweb/+/38570
Implement LocationIndicatorActive for CPU resource | https://gerrit.openbmc.org/c/openbmc/bmcweb/+/37045

For the HWIsolation, the commits do not have respective upstream commits, as we do not have a standard guard implementation in the upstream

Test output:
After adding the patches, the Hardware Isolation records are 
-bash-4.2$ curl -k -H "X-Auth-Token: $bmc_token" -XGET https://${BMC_IP}/redfish/v1/Systems/system/LogServices/HardwareIsolation/Entries
{
  "@odata.id": "/redfish/v1/Systems/system/LogServices/HardwareIsolation/Entries",
  "@odata.type": "#LogEntryCollection.LogEntryCollection",
  "Description": "Collection of System Hardware Isolation Entries",
  "Members": [
    {
      "@odata.id": "/redfish/v1/Systems/system/LogServices/HardwareIsolation/Entries/1",
      "@odata.type": "#LogEntry.v1_9_0.LogEntry",
      "Created": "2023-02-12T15:11:24+00:00",
      "EntryType": "Event",
      "Id": "1",
      "Links": {
        "OriginOfCondition": {
          "@odata.id": "/redfish/v1/Systems/system/Memory/dimm2"
        }
      },
      "Message": "DDR Memory Port",
      "Name": "Hardware Isolation Entry",
      "Severity": "Warning"
    },

gtmills commented:
What is the upstream commits here? For next time, "[sync_1030_commits](https://github.com/deepakala-k/bmcweb/tree/sync_1030_commits)" doesn't tell me about what the branch does (we are syncing all our commits), a branch name like 1050-hardware-isolation would be a lot more useful. Can you please edit the title to include the feature, hardware isolation ?

gtmills commented:
This is such as massive PR, 
can we start with 2 PRs, one for pretty name and one for SubProcessors:

PR 1)
[util: Add pretty name for resources](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/3388298927fbde327399c8d87c9fd4a5a3efa635)
wltu authored and deepakala-k committed Feb 14, 2023

PR 2) 
[Implement LocationIndicatorActive for CPU resource](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/e7fe758a77fb2f63acdd79ea705e0a9220163d89)
lxwinspur authored and deepakala-k committed Feb 14, 2023

[Implement SubProcessors for processor core](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/c8b654aec12ca2ba2c60e9e46dc669c457fb01d9)

deepakala-k commented:
> What is the upstream commits here? For next time, "[sync_1030_commits](https://github.com/deepakala-k/bmcweb/tree/sync_1030_commits)" doesn't tell me about what the branch does (we are syncing all our commits), a branch name like 1050-hardware-isolation would be a lot more useful. Can you please edit the title to include the feature, hardware isolation ?

Here is the upstream commits for the relevant ones. 

Implement LocationIndicatorActive for CPU resource | https://gerrit.openbmc.org/c/openbmc/bmcweb/+/37045

Implement SubProcessors for processor core | https://gerrit.openbmc.org/c/openbmc/bmcweb/+/38570

Implement LocationIndicatorActive for CPU resource | https://gerrit.openbmc.org/c/openbmc/bmcweb/+/37045

All the other commits do not have upstream commits as we do not have a standard guard implementation in the upstream.

deepakala-k commented:
> This is such as massive PR, can we start with 2 PRs, one for pretty name and one for SubProcessors:
> 
> PR 1) [util: Add pretty name for resources](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/3388298927fbde327399c8d87c9fd4a5a3efa635) wltu authored and deepakala-k committed Feb 14, 2023
> 
> PR 2) [Implement LocationIndicatorActive for CPU resource](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/e7fe758a77fb2f63acdd79ea705e0a9220163d89) lxwinspur authored and deepakala-k committed Feb 14, 2023
> 
> [Implement SubProcessors for processor core](https://github.com/ibm-openbmc/bmcweb/pull/542/commits/c8b654aec12ca2ba2c60e9e46dc669c457fb01d9)

I had to include hardware isolation commits and other related commits under a single pull request (PR) because they were interconnected.
During this process, I also had to update a portion of the code that had been added as part of another commit.
Ignoring these updates was not an option, as the continuous integration (CI) process would fail without them.
To provide clarity, I organized the commits into groups, even though they were not originally added in the same order, and pulled them whenever dependencies were identified.
Please find the list of commits based on the feature

HWIsolation feature related commits:
1. util: Add pretty name for resources - 3388298927fbde327399c8d87c9fd4a5a3efa635
2. redfish-core: LogServices: Added HardwareIsolation service  - 8e0292bb4e5f813cb733682fe1ff4382c5ccb704
3. LogServices: HardwareIsolation: Get LogEntryCollection - 2d4632e8f22e971e20653397ec947b0631d0dd8d
4. LogServices: HardwareIsolation: Get LogEntry - 3345042d20404bf8c6f985b505ffbfaff464acbe
5. LogServices: HardwareIsolation: Delete LogEntry  - 3338f71008b8f25032e058df240d0a4f365ca0ef
6. LogServices: HardwareIsolation: Post ClearLog - 76392d7fe291f5134e2a88805d52e95cdc5ab625
7. redfish-core: Core: Enabled the isolation (aka guard) feature  - e332321f7b0c5f0cc4fe4370e63a0bf326a3f2c8
8. redfish-core: Memory: Enabled the isolation (aka guard) feature - f338882aba0019cdd48a1e92b903fdcb290cadd2
9. redfish-core: Processor: Fixed the processor object search (ibm-openbmc#168) - cf597f0b7016dbf59e078dabbeadd535107958eb
10. Enabled deconfiguration reason support to the DIMM and Core and Few fixes (ibm-openbmc#171) - 5b79cb49ce2c3955bed52210c07f88a72e66913c
11. HW-Isolation: Fix, Use GetAncestors to get the parents id (ibm-openbmc#235) - 079c6076160833f00bfe4906af035564f66464bf
12. HW-Isolation: Fix, Don't throw internal error if failed to get error log (ibm-openbmc#245) - 028473788445e5e995c334885dd7f6c02cffd2c8
13. Core: Fix, Patch a core into the respective parent processor (ibm-openbmc#261) - 54c17681a421b37e892b1d9a04055334e2c96c08
14. HW-Isolation: Fix, Update State if the Core and DIMM are recovered (ibm-openbmc#288) - 2c7415503265b618e9868f72546647b006b3618d
15. HW-Isolation: Return an appropriate error if the request is failed. - a037b82e1d82607d72c50b8f2104e7a8c71e569a
16. HW-Isolation: Fill OriginOfCondition for the TPM and Motherboard (ibm-openbmc#278) - c2dbb83537e57802164a53c65f3424b99b2d52df
17. HW-Isolation: Return ResourceCannotBeDeleted error (ibm-openbmc#297)  - 438667b85e7b218a529102b2e3cc6e1778b06380
18. HW-Isolation: Fix, Don't handle the Unavailable D-Bus error - 3ced7d174cbdb32a8b73fe11c5d0b137832499be
19. LogService: HW-Isolation: Return an appropriate error - a7d37e519e28c803baae05d25917451fa30d02ba
20. LogEntry: HW-Isolation: Removed the Resolved property (ibm-openbmc#341) - 54f3f0ed567682b13cd11ea5529d74607665b84f
21. Add missing odata.id field under OriginOfCondition for assemblies (ibm-openbmc#494) - 6e065a5d08b4f0c7ded01c08daf0ed48ab8d47e3

Other feature/dependent commits:
1. Implement LocationIndicatorActive for CPU resource - e7fe758a77fb2f63acdd79ea705e0a9220163d89
2. Implement SubProcessors for processor core - c8b654aec12ca2ba2c60e9e46dc669c457fb01d9
3. Implement LocationIndicatorActive for Memory resource - b45cd3209c9b047321c264354d36be798d506f68
4. redfish-core: Processor: Workaround to handle DCM - b561af9ed5451aee9dd93961495a560ad090ae2b
5. registry: Add PropertyValueResourceConflict registry - d9d3f1944a93a6aba77b9fab20b0dd232dc0e769
6. registry: Add PropertyValueExternalConflict registry - e30cd0482e8bbb8f0dded1c66d88c72d12a8318e

Common commits that addresses issues from all the other commits
1. clang-format ran - 32b0877754af0f964f93bae1bd0fd8fa45a57afd
2. Fix Errors found during CI - 8fee226751dafccd5060ab38aa1fb314d3bc3eda

RameshIyyar reviewed (approved):
IMO, this is a big list of commits that need to pull from 1030 to 1050 to enable hardware isolation service (as mentioned early we cannot upstream it until some better standard design as discussed in the past) and all are dependent on each other so I think we won't get any benefits by creating multiple PRs, and even If we tried to create multiple PR then, we struggle to meet to time to deliver 1050 driver because we will see a lot of conflicts and error as mentioned in the last 2 commits so I would like to go with single PR that also helps in a different angle from the future perspective that is, we can refer this PR as a reference to migrating upstream again.

Nothing running my head on this to give some better options. Please feel free to share options that it makes easy for us to get a lot of commits.

Thanks @deepakala-k,  I noticed a few things early and it seems you addressed them in the latest patch set.

gtmills commented on redfish-core/include/error_messages.hpp:
```diff
@@ -482,6 +482,41 @@ nlohmann::json propertyValueIncorrect(std::string_view arg1,
+ * @param[in] arg3 Parameter of message that will replace %3 in its body.
+ *
+ * @returns Message PropertyValueResourceConflict to JSON */
+nlohmann::json propertyValueResourceConflict(const std::string& arg1,
```
Why can't this go upstream?

gtmills replied:
And the rest of these in this file..

RameshIyyar replied:
When we discussed early on this, I heard that we cannot push the registry without any use in the upstream bmcweb that is, we are only using it for the hw-isolation so its stays in the downstream

gtmills replied:
> When we discussed early on this, I heard that we cannot push the registry without any use in the upstream bmcweb that is, we are only using it for the hw-isolation so its stays in the downstream


Why, can't we push to the registry ? I don't see a problem with that, hopefully one day this will be upstream

RameshIyyar replied:
Even I am not seeing any problem pushing registry-specific changes upstream (remembered like some process in the past so if something changed I am happy with it :)) - IMO, we can try that in parallel if no restriction now.

gtmills commented:
The number of lines changed here is huge. 
The number of files changed here is a lot (16). A lot of these changes are to core redfish-core files, these will not be easy to maintain downstream.. I really think we should look at upstreaming gard/deconfigure, a feature HPE and Dell both have. I really think this could go upstream if we work with the community.

 I think this makes rebasing 1050 really hard. :( 
 
 Can we look at getting as much of this upstream as possible?

gtmills commented on redfish-core/lib/log_services.hpp:
```diff
@@ -5368,4 +5419,1084 @@ inline void requestRoutesPostCodesEntry(App& app)
+        .privileges(redfish::privileges::getLogService)
+        .methods(boost::beast::http::verb::get)(
+            getSystemHardwareIsolationLogService);
+
```
New bmcweb code upstream also has HEAD. 
Oh well.

deepakala-k replied:
I am unable to follow the comment. Could you please elaborate it.
Thank you!

gtmills commented on redfish-core/lib/memory.hpp:
```diff
@@ -26,11 +26,30 @@
 {
 
+using DimmProperty =
+    std::variant<std::string, std::vector<uint32_t>, std::vector<uint16_t>,
```
include what you use?
#include <variant>
#include <string>
#include <vector>

deepakala-k replied:
Sure

deepakala-k replied:
Thank you for your valid comments. We have created the JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 to address this comment.
The scope of this task was to cherry pick the working HWIsolation commits from 1030. And as the number of commits are more, it is difficult to refactor them to align with upstream as part of this PR. So we will address all them as a separate story.

gtmills commented on redfish-core/include/utils/error_log_utils.hpp:
```diff
@@ -0,0 +1,84 @@
+#pragma once
+
```
include what you use

gtmills replied:
Can you run https://include-what-you-use.org/ ?

deepakala-k replied:
Thank you for your valid comments. We have created the JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 to address this comment.
The scope of this task was to cherry pick the working HWIsolation commits from 1030. And as the number of commits are more, it is difficult to refactor them to align with upstream as part of this PR. So we will address all them as a separate story.

gtmills commented on redfish-core/lib/memory.hpp:
```diff
@@ -698,70 +722,308 @@ inline void getDimmPartitionData(std::shared_ptr<bmcweb::AsyncResp> aResp,
+            }
+        }
+        },
+        service, path, "org.freedesktop.DBus.Properties", "GetAll",
```
We use     sdbusplus::asio::getAllProperties( upstream like above on line 628

gtmills commented on redfish-core/include/utils/error_log_utils.hpp:
```diff
@@ -0,0 +1,84 @@
+        }
+        },
+        "xyz.openbmc_project.Logging", errorLogObjPath.str,
+        "org.freedesktop.DBus.Properties", "Get",
```
We are using     sdbusplus::asio::getProperty( upstream

deepakala-k replied:
Thank you for your valid comments. We have created the JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 to address this comment.
The scope of this task was to cherry pick the working HWIsolation commits from 1030. And as the number of commits are more, it is difficult to refactor them to align with upstream as part of this PR. So we will address all them as a separate story.

gtmills commented on redfish-core/include/utils/hw_isolation.hpp:
```diff
@@ -0,0 +1,730 @@
+            std::array<const char*, 1>{"xyz.openbmc_project.Logging.Event"});
+        },
+        "xyz.openbmc_project.ObjectMapper", resourceObjPath.str + "/event_log",
+        "org.freedesktop.DBus.Properties", "Get",
```
sdbusplus::asio::getProperty

deepakala-k replied:
Thank you for your valid comments. We have created the JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 to address this comment.
The scope of this task was to cherry pick the working HWIsolation commits from 1030. And as the number of commits are more, it is difficult to refactor them to align with upstream as part of this PR. So we will address all them as a separate story.

gtmills commented on redfish-core/lib/assembly.hpp:
```diff
@@ -532,6 +538,215 @@ inline void getChassis(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
+inline void fillWithAssemblyId(
+    const std::shared_ptr<bmcweb::AsyncResp>& aResp,
+    const std::string& assemblyParentServ,
+    const sdbusplus::message::object_path& assemblyParentObjPath,
```
include what you use ?

deepakala-k replied:
Thank you for your valid comments. We have created the JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 to address this comment.
The scope of this task was to cherry pick the working HWIsolation commits from 1030. And as the number of commits are more, it is difficult to refactor them to align with upstream as part of this PR. So we will address all them as a separate story.

gtmills commented on redfish-core/lib/memory.hpp:
```diff
@@ -820,7 +1082,32 @@ inline void requestRoutesMemory(App& app)
+        });
+
+    BMCWEB_ROUTE(app, "/redfish/v1/Systems/system/Memory/<str>/")
+        .privileges({{"Login"}})
```
let's use the privilege framework 

redfish::privileges::getMemory

deepakala-k replied:
Thank you for your valid comments. We have created the JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 to address this comment.
The scope of this task was to cherry pick the working HWIsolation commits from 1030. And as the number of commits are more, it is difficult to refactor them to align with upstream as part of this PR. So we will address all them as a separate story.

deepakala-k commented:
> The number of lines changed here is huge. The number of files changed here is a lot (16). A lot of these changes are to core redfish-core files, these will not be easy to maintain downstream.. I really think we should look at upstreaming gard/deconfigure, a feature HPE and Dell both have. I really think this could go upstream if we work with the community.
> 
> I think this makes rebasing 1050 really hard. :(
> 
> Can we look at getting as much of this upstream as possible?


Sure. We will bring it up in our team meetings and try to come up with a plan.

deepakala-k commented:
> @deepakala-k Did the Redfish Validator pass?

Thank you for the pointer. The redfish validator does not pass. I am trying to fix it. Will get it done ASAP and share the link here

gtmills commented on redfish-core/lib/log_services.hpp:
```diff
@@ -5368,4 +5419,1084 @@ inline void requestRoutesPostCodesEntry(App& app)
+                    uriPropPath /= entryJsonIdx - 1;
+                    uriPropPath /= "Links";
+                    uriPropPath /= "OriginOfCondition";
+
```
This needs odata.id like we have today: 
https://github.com/ibm-openbmc/bmcweb/blob/1030/redfish-core/lib/log_services.hpp#L4799

deepakala-k replied:
Thank you! Added to the refactor JIRA ticket https://jsw.ibm.com/browse/PFEBMC-620 and will be addressed as part of it

gtmills replied:
we should address this here, otherwise we have a validator error

deepakala-k replied:
Addressed it. Here is the link to redfish validator results.
https://sys-openbmc-dev-jenkins.swg-devops.com/job/OpenBMC-Dev/job/test/job/common/job/Redfish-Service-Validator/18118/artifact/logs/ConformanceHtmlLog_02_20_2023_133127.html

deepakala-k commented:
> > @deepakala-k Did the Redfish Validator pass?
> 
> Thank you for the pointer. The redfish validator does not pass. I am trying to fix it. Will get it done ASAP and share the link here


Redfish validator tests run WITHOUT bmcweb HWIsolation changes:
https://sys-openbmc-dev-jenkins.swg-devops.com/job/OpenBMC-Dev/job/test/job/common/job/Redfish-Service-Validator/18098/artifact/logs/ConformanceHtmlLog_02_17_2023_065203.html

Redfish validator tests run WITH bmcweb HWIsolation changes:
https://sys-openbmc-dev-jenkins.swg-devops.com/job/OpenBMC-Dev/job/test/job/common/job/Redfish-Service-Validator/18093/artifact/logs/ConformanceHtmlLog_02_16_2023_124337.html

The assembly related failure that is introduced as we are trying to access more inventory items. I have created a separate JIRA ticket to track and close that item. https://jsw.ibm.com/browse/PFEBMC-621

gtmills commented on redfish-core/lib/update_service.hpp:
```diff
@@ -1065,6 +1066,10 @@ inline void requestRoutesSoftwareInventory(App& app)
                                    *swId);
+                asyncResp->res.jsonValue["Name"] = "Software Inventory";
+
+                name_util::getPrettyName(asyncResp, obj.first, obj.second,
```
We don't want this here: https://github.com/ibm-openbmc/bmcweb/commit/2ff339a970456d637836b5202d8b13cf1a06d0b9
